File: JavaScript/5-BinarySearch/Problems/1-BinarySearch/BinarySearch.js
// Binary Search - 5 Different Approaches

// 1. Iterative Binary Search
function binarySearchIterative(arr, target) {
  let low = 0;
  let high = arr.length - 1;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);

    if (arr[mid] === target) {
      return mid;
    } else if (arr[mid] < target) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return -1;
}

// 2. Recursive Binary Search
function binarySearchRecursive(arr, target, low, high) {
  if (low > high) {
    return -1;
  }

  const mid = Math.floor((low + high) / 2);

  if (arr[mid] === target) {
    return mid;
  } else if (arr[mid] < target) {
    return binarySearchRecursive(arr, target, mid + 1, high);
  } else {
    return binarySearchRecursive(arr, target, low, mid - 1);
  }
}

// 3. Binary Search with Early Exit
function binarySearchEarlyExit(arr, target) {
    let low = 0;
    let high = arr.length - 1;

    while (low < high) { // Changed condition for early exit
        const mid = Math.floor((low + high) / 2);

        if (arr[mid] === target) {
            return mid;
        } else if (arr[mid] < target) {
            low = mid + 1;
        } else {
            high = mid; // Changed to high = mid
        }
    }
    //check the last element
    if(arr[low] === target){
        return low;
    }
    return -1;
}

// 4. Binary Search using Array.prototype.slice (Less Efficient)
function binarySearchSlice(arr, target) {
  let low = 0;
  let high = arr.length - 1;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const midValue = arr[mid];

    if (midValue === target) {
      return mid;
    } else if (midValue < target) {
      low = mid + 1;
      arr = arr.slice(low, high + 1); // Creating a new sub-array
      high = arr.length -1;
      low = 0;
    } else {
      high = mid - 1;
      arr = arr.slice(low, high + 1);  // Creating a new sub-array
      high = arr.length -1;
      low = 0;
    }
  }
  return -1;
}

// 5. Binary Search with ES6 findIndex (Not a true binary search, but uses findIndex)
function binarySearchFindIndex(arr, target) {
  const index = arr.findIndex(item => item === target);
  return index; // findIndex returns -1 if not found
}



// Example Usage:
const sortedArray = [-1, 0, 3, 5, 9, 12];
const targetValue = 9;

console.log("Array: ", sortedArray);
console.log("Target: ", targetValue);
console.log("\n--- Binary Search Results ---");

// 1. Iterative
const result1 = binarySearchIterative(sortedArray, targetValue);
console.log("1. Iterative:", result1); // Output: 4

// 2. Recursive
const result2 = binarySearchRecursive(sortedArray, targetValue, 0, sortedArray.length - 1);
console.log("2. Recursive:", result2); // Output: 4

// 3. Early Exit
const result3 = binarySearchEarlyExit(sortedArray, targetValue);
console.log("3. Early Exit:", result3); // Output: 4

// 4. Using Slice (Less Efficient)
const result4 = binarySearchSlice(sortedArray, targetValue);
console.log("4. Using Slice:", result4); // Output: 4

// 5. Using findIndex (Not a true binary search)
const result5 = binarySearchFindIndex(sortedArray, targetValue);
console.log("5. Using findIndex:", result5); // Output: 4

// Example where target is not found
const targetNotFound = 2;
console.log("\nTarget Not Found Example:");
console.log("Target:", targetNotFound);
console.log("Iterative:", binarySearchIterative(sortedArray, targetNotFound)); // Output: -1
console.log("Recursive:", binarySearchRecursive(sortedArray, targetNotFound, 0, sortedArray.length - 1)); // Output: -1
console.log("Early Exit:", binarySearchEarlyExit(sortedArray, targetNotFound));
console.log("Slice:", binarySearchSlice(sortedArray, targetNotFound));
console.log("findIndex:", binarySearchFindIndex(sortedArray, targetNotFound));

File: C++/5-BinarySearch/Problems/1-BinarySearch/BinarySearch.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

// Binary Search - Iterative Approach 1
// Uses a standard iterative approach with integer mid calculation.
int binarySearchIterative1(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2; // Prevent potential overflow
        if (arr[mid] == target) {
            return mid; // Target found, return the index
        } else if (arr[mid] < target) {
            left = mid + 1; // Target is in the right half
        } else {
            right = mid - 1; // Target is in the left half
        }
    }
    return -1; // Target not found
}

// Binary Search - Iterative Approach 2
// Similar to the first iterative approach, but with a slight variation in the loop condition.
int binarySearchIterative2(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while (left < right) { // Loop continues as long as there are at least two elements to check
        int mid = left + (right - left) / 2;
        if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid; // Important: right is set to mid, not mid-1
        }
    }
    // After the loop, left == right, and it's the potential target
    if (arr[left] == target) {
        return left;
    }
    return -1;
}


// Binary Search - Recursive Approach 1
// A straightforward recursive implementation.
int binarySearchRecursive1Helper(const std::vector<int>& arr, int target, int left, int right) {
    if (left > right) {
        return -1; // Base case: target not found
    }
    int mid = left + (right - left) / 2;
    if (arr[mid] == target) {
        return mid;
    } else if (arr[mid] < target) {
        return binarySearchRecursive1Helper(arr, target, mid + 1, right); // Recursive call on right half
    } else {
        return binarySearchRecursive1Helper(arr, target, left, mid - 1); // Recursive call on left half
    }
}

int binarySearchRecursive1(const std::vector<int>& arr, int target) {
    return binarySearchRecursive1Helper(arr, target, 0, arr.size() - 1);
}

// Binary Search - Recursive Approach 2
// Tail-recursive version (often optimized by compilers).  Less common, but good to know.
int binarySearchRecursive2Helper(const std::vector<int>& arr, int target, int left, int right, int& result) {
    if (left > right) {
        return -1; // Base case: target not found.  Return value doesn't matter here, result is used.
    }
    int mid = left + (right - left) / 2;
    if (arr[mid] == target) {
        result = mid; // Store the result
        return mid; // early return
    } else if (arr[mid] < target) {
        return binarySearchRecursive2Helper(arr, target, mid + 1, right, result);
    } else {
        return binarySearchRecursive2Helper(arr, target, left, mid - 1, result);
    }
}

int binarySearchRecursive2(const std::vector<int>& arr, int target) {
    int result = -1;
    binarySearchRecursive2Helper(arr, target, 0, arr.size() - 1, result);
    return result;
}

// Binary Search - Using STL (std::lower_bound)
// Demonstrates using the C++ Standard Template Library's lower_bound function
// which performs a binary search.  Returns an iterator.
int binarySearchSTL(const std::vector<int>& arr, int target) {
    auto it = std::lower_bound(arr.begin(), arr.end(), target);
    if (it != arr.end() && *it == target) {
        return std::distance(arr.begin(), it); // Calculate the index
    }
    return -1;
}

int main() {
    std::vector<int> sortedArray = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};
    int targetValue = 23;

    std::cout << "Array: ";
    for (int num : sortedArray) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
    std::cout << "Target: " << targetValue << std::endl;

    // Test each binary search implementation
    int index1 = binarySearchIterative1(sortedArray, targetValue);
    std::cout << "Iterative 1: Index of " << targetValue << " is " << index1 << std::endl;

    int index2 = binarySearchIterative2(sortedArray, targetValue);
    std::cout << "Iterative 2: Index of " << targetValue << " is " << index2 << std::endl;

    int index3 = binarySearchRecursive1(sortedArray, targetValue);
    std::cout << "Recursive 1: Index of " << targetValue << " is " << index3 << std::endl;

    int index4 = binarySearchRecursive2(sortedArray, targetValue);
    std::cout << "Recursive 2: Index of " << targetValue << " is " << index4 << std::endl;

    int index5 = binarySearchSTL(sortedArray, targetValue);
    std::cout << "STL: Index of " << targetValue << " is " << index5 << std::endl;

    // Test case where the target is not in the array
    targetValue = 100;
    std::cout << "\nTarget: " << targetValue << std::endl;
    index1 = binarySearchIterative1(sortedArray, targetValue);
    std::cout << "Iterative 1: Index of " << targetValue << " is " << index1 << std::endl;

    index2 = binarySearchIterative2(sortedArray, targetValue);
    std::cout << "Iterative 2: Index of " << targetValue << " is " << index2 << std::endl;

    index3 = binarySearchRecursive1(sortedArray, targetValue);
    std::cout << "Recursive 1: Index of " << targetValue << " is " << index3 << std::endl;

    index4 = binarySearchRecursive2(sortedArray, targetValue);
    std::cout << "Recursive 2: Index of " << targetValue << " is " << index4 << std::endl;

    index5 = binarySearchSTL(sortedArray, targetValue);
    std::cout << "STL: Index of " << targetValue << " is " << index5 << std::endl;

    return 0;
}
